examples/capture/main.c

```diff
@@ -0,0 +1,196 @@
+
+void request_adapter_callback(WGPUAdapterId received, void *userdata) {
+    *(WGPUAdapterId *) userdata = received;
+}
+
+void read_buffer_map(
+        WGPUBufferMapAsyncStatus status,
+        uint8_t *userdata) {
+}
+
+typedef struct BufferDimensions {
+    uint32_t width;
+    uint32_t height;
+    uint32_t unpadded_bytes_per_row;
+    uint32_t padded_bytes_per_row;
+} BufferDimensions;
+
+BufferDimensions buffer_dimensions_new(int width, int height) {
+    int bytes_per_pixel = 4;
+    int unpadded_bytes_per_row = width * bytes_per_pixel;
+    int align = 256;
+    int padded_bytes_per_row_padding = (align - unpadded_bytes_per_row % align) % align;
+    int padded_bytes_per_row = unpadded_bytes_per_row + padded_bytes_per_row_padding;
+    return (BufferDimensions) {
+            .width = width,
+            .height = height,
+            .padded_bytes_per_row = padded_bytes_per_row,
+            .unpadded_bytes_per_row = unpadded_bytes_per_row
+    };
+}
+
+WGPULimits wgpu_limits_default() {
+    return (WGPULimits) {
+            .max_bind_groups = 4
+    };
+}
+
+/*
+ * No automatic defaults generated, have to generate the struct's defaults by hand.
+ */
+WGPUTextureViewDescriptor wgpu_texture_view_descriptor_default() {
+    return (WGPUTextureViewDescriptor) {
+            .label = NULL,
+            .format = 90, // see https://github.com/gfx-rs/wgpu-native/issues/67
+            .dimension = 6, // same as above
+            .aspect = WGPUTextureAspect_All, // non-nullable, can use 0th element
+            .array_layer_count = 0, // 0 = absent in NonZero ints
+            .base_array_layer = 0,
+            .base_mip_level = 0,
+            .level_count = 0 // 0 = absent in NonZero ints
+    };
+}
+
+int main(
+        int argc,
+        char *argv[]) {
+
+    int width = 100;
+    int height = 200;
+
+    WGPUAdapterId adapter = {0};
+    wgpu_request_adapter_async(
+            &(WGPURequestAdapterOptions) {
+                    .power_preference = WGPUPowerPreference_HighPerformance,
+                    .compatible_surface = 0
+            },
+            2 | 4 | 8,
+            request_adapter_callback,
+            (void *) &adapter
+    );
+
+    WGPUDeviceId device = wgpu_adapter_request_device(adapter,
+                                                      &(WGPUDeviceDescriptor) {
+                                                              .label = NULL,
+                                                              0,
+                                                              wgpu_limits_default(),
+                                                              NULL}
+    );
+
+    BufferDimensions buffer_dimensions = buffer_dimensions_new(width, height);
+    WGPUBufferId output_buffer = wgpu_device_create_buffer(device,
+                                                           &(WGPUBufferDescriptor) {
+                                                                   .label = NULL,
```

Comment:
could we use nicer formatting for C code?

Reply:
Yeah sorry about that, CLion has wild default formatting rules. I've used additional newlines to force better formatting.